TYP: issues with type-checking against numpy 2.1

Type-checking is now failing against NumPy 2.1 ([example logs](https://github.com/1313e/CMasher/actions/runs/10650551220/job/29522078891))

We're seeing two errors

```
cmasher/utils.py: note: In function "_get_cmap_lightness_rank":
cmasher/utils.py:167: error: Unsupported left operand type for - (Never)  [operator]
cmasher/utils.py: note: At top level:
cmasher/utils.py:1554: error: Unused "type: ignore" comment  [unused-ignore]
Found 2 errors in 1 file (checked 63 source files)
```

The first one I think is a bug in NumPy, which I'll report upstream if it hasn't been done already.
The second one (`Unused "type: ignore" comment`) would be easy to solve (remove the unused comment), but unfortunately the comment is still needed on Python 3.9, at it isn't supported by NumPy 2.1. Fixing this part will require either dropping Python 3.9 (which is maybe a bit radical, unless we see it as a tradeoff for adding 3.13 (which I intend to do very soon)), or temporarily disabling this warning in mypy's configuration (which sounds much more reasonable).

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -25,7 +25,6 @@ jobs:
         - macos-latest
         - windows-latest
         python-version:
-        - '3.9'
         - '3.10'
         - '3.11'
         - '3.12'
@@ -34,9 +33,8 @@ jobs:
 
         include:
         - os: ubuntu-20.04
-          python-version: '3.9'
+          python-version: '3.10'
           deps: minimal
-          pip-args: --constraint requirements/min_constraints.txt
           with-viscm: false
 
       fail-fast: false
@@ -63,9 +61,11 @@ jobs:
       run: uv pip install "viscm>=0.10"
 
     - if: matrix.deps == 'minimal'
+      # 2021-11-16 is matplotlib 3.5.0's release date
       run: |
-        uv pip compile pyproject.toml --resolution=lowest-direct | uv pip install -r - \
-          --only-binary ':all:' ${{ matrix.pip-args }}
+        uv pip compile pyproject.toml \
+         --resolution=lowest-direct \
+         --exclude-newer '2021-11-16' | uv pip install --no-build -r -
 
     - name: Build
       run: |
@@ -86,7 +86,7 @@ jobs:
     strategy:
       matrix:
         python-version:
-        - '3.9'
+        - '3.10'
         - '3.12'
 
     runs-on: ubuntu-latest
@@ -127,7 +127,7 @@ jobs:
     - name: Setup Python
       uses: actions/setup-python@v5
       with:
-        python-version: '3.9'
+        python-version: '3.12'
     - uses: yezz123/setup-uv@v4
       with:
         uv-version: 0.2.18

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -26,13 +26,6 @@ repos:
   - id: codespell
     args: [--ignore-words-list, 'mutch, ore']
 
-- repo: https://github.com/PyCQA/bandit
-  rev: 1.7.9
-  hooks:
-  - id: bandit
-    # B101 is assert statements
-    args: [--recursive, --skip, B101, .]
-
 - repo: https://github.com/macisamuele/language-formatters-pre-commit-hooks
   rev: v2.13.0
   hooks:

--- a/.readthedocs.yaml
+++ b/.readthedocs.yaml
@@ -8,7 +8,7 @@ version: 2
 build:
   os: ubuntu-22.04
   tools:
-    python: '3.9'
+    python: '3.12'
 
 # Build documentation in the "docs/" directory with Sphinx
 sphinx:

--- a/cmasher/utils.py
+++ b/cmasher/utils.py
@@ -9,11 +9,12 @@ Utility functions for registering and manipulating colormaps in various ways.
 # Built-in imports
 import sys
 from collections import OrderedDict
+from collections.abc import Callable
 from glob import glob
 from importlib.util import find_spec
 from os import path
 from textwrap import dedent
-from typing import TYPE_CHECKING, Callable, Optional, Union
+from typing import TYPE_CHECKING, Optional, Union
 
 import matplotlib as mpl
 import numpy as np
@@ -34,12 +35,9 @@ from cmasher import cm as cmrcm
 from ._known_cmap_types import _CMASHER_BUILTIN_MAP_TYPES
 
 if TYPE_CHECKING:
-    from matplotlib.artist import Artist
+    from typing import TypeAlias
 
-    if sys.version_info >= (3, 10):
-        from typing import TypeAlias
-    else:
-        from typing_extensions import TypeAlias
+    from matplotlib.artist import Artist
 
 _HAS_VISCM = find_spec("viscm") is not None
 
@@ -65,7 +63,7 @@ __all__ = [
 cspace_convert = cspace_converter("sRGB1", "CAM02-UCS")
 
 # Type aliases
-CMAP = Union[str, Colormap]
+CMAP = str | Colormap
 RED: "TypeAlias" = float
 GREEN: "TypeAlias" = float
 BLUE: "TypeAlias" = float
@@ -241,8 +239,8 @@ def _get_cmap_perceptual_rank(
 # %% FUNCTIONS
 # This function combines multiple colormaps at given nodes
 def combine_cmaps(
-    *cmaps: Union[Colormap, str],
-    nodes: Optional[Union[list[float], np.ndarray]] = None,
+    *cmaps: Colormap | str,
+    nodes: list[float] | np.ndarray | None = None,
     n_rgb_levels: int = 256,
     combined_cmap_name: str = "combined_cmap",
 ) -> LinearSegmentedColormap:
@@ -297,7 +295,7 @@ def combine_cmaps(
     if len(cmaps) <= 1:
         raise ValueError("Expected at least two colormaps to combine.")
     for cm in cmaps:
-        if not isinstance(cm, (Colormap, str)):
+        if not isinstance(cm, Colormap | str):
             raise TypeError(f"Unsupported colormap type: {type(cm)}.")
     _cmaps: list[Colormap] = [
         cm if isinstance(cm, Colormap) else mpl.colormaps[cm] for cm in cmaps
@@ -306,7 +304,7 @@ def combine_cmaps(
     # Generate default nodes for equal separation
     if nodes is None:
         nodes_arr = np.linspace(0, 1, len(_cmaps) + 1)
-    elif isinstance(nodes, (list, np.ndarray)):
+    elif isinstance(nodes, list | np.ndarray):
         nodes_arr = np.concatenate([[0.0], nodes, [1.0]])
     else:
         raise TypeError(f"Unsupported nodes type: {type(nodes)}, expect list of float.")
@@ -344,7 +342,7 @@ def combine_cmaps(
 
 # This function creates a standalone module of a CMasher colormap
 def create_cmap_mod(
-    cmap: str, *, save_dir: str = ".", _copy_name: Optional[str] = None
+    cmap: str, *, save_dir: str = ".", _copy_name: str | None = None
 ) -> str:
     """
     Creates a standalone Python module of the provided *CMasher* `cmap` and
@@ -484,16 +482,16 @@ def create_cmap_mod(
 
 # This function creates an overview plot of all colormaps specified
 def create_cmap_overview(
-    cmaps: Union[list[CMAP], dict[str, list[Colormap]], None] = None,
+    cmaps: list[CMAP] | dict[str, list[Colormap]] | None = None,
     *,
-    savefig: Optional[str] = None,
+    savefig: str | None = None,
     use_types: bool = True,
-    sort: Optional[Union[str, Callable]] = "alphabetical",
+    sort: str | Callable | None = "alphabetical",
     show_grayscale: bool = True,
     show_info: bool = False,
-    plot_profile: Union[bool, float] = False,
+    plot_profile: bool | float = False,
     dark_mode: bool = False,
-    title: Optional[str] = "Colormap Overview",
+    title: str | None = "Colormap Overview",
     wscale: float = 1,
     hscale: float = 1,
 ) -> None:
@@ -604,7 +602,7 @@ def create_cmap_overview(
             )
 
     # Create empty list of cmaps
-    cmaps_list: list[Union[Colormap, tuple[str, bool]]] = []
+    cmaps_list: list[Colormap | tuple[str, bool]] = []
 
     # Define empty dict of colormaps
     cmaps_dict: OrderedDict[str, list[Colormap]] = OrderedDict()
@@ -749,7 +747,7 @@ def create_cmap_overview(
         axs = np.array([axs])
 
     # Loop over all cmaps defined in cmaps list
-    for ax, _cm in zip(axs, cmaps_list):
+    for ax, _cm in zip(axs, cmaps_list, strict=True):
         # Obtain axes objects and turn them off
         if show_grayscale:
             # Obtain Axes objects
@@ -841,7 +839,9 @@ def create_cmap_overview(
                     segments = np.split(points, s_idx)
 
                     # Loop over all pairs of adjacent segments
-                    for i, (seg1, seg2) in enumerate(zip(segments[:-1], segments[1:])):
+                    for i, (seg1, seg2) in enumerate(
+                        zip(segments[:-1], segments[1:], strict=True)
+                    ):
                         # Determine the point in the center of these segments
                         central_point = (seg1[-1] + seg2[0]) / 2
 
@@ -1085,9 +1085,7 @@ def get_cmap_type(cmap: CMAP) -> str:
 
 
 # Function create a colormap using a subset of the colors in an existing one
-def get_sub_cmap(
-    cmap: CMAP, start: float, stop: float, *, N: Optional[int] = None
-) -> LC:
+def get_sub_cmap(cmap: CMAP, start: float, stop: float, *, N: int | None = None) -> LC:
     """
     Creates a :obj:`~matplotlib.cm.ListedColormap` object using the colors in
     the range `[start, stop]` of the provided `cmap` and returns it.
@@ -1448,7 +1446,7 @@ def set_cmap_legend_entry(artist: "Artist", label: str) -> None:
 # Function to take N equally spaced colors from a colormap
 def take_cmap_colors(
     cmap: CMAP,
-    N: Optional[int],
+    N: int | None,
     *,
     cmap_range: tuple[float, float] = (0, 1),
     return_fmt: str = "float",
@@ -1565,7 +1563,7 @@ def take_cmap_colors(
 def view_cmap(
     cmap: CMAP,
     *,
-    savefig: Optional[str] = None,
+    savefig: str | None = None,
     show_test: bool = False,
     show_grayscale: bool = False,
 ) -> None:

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -19,7 +19,6 @@ classifiers = [
     "Operating System :: Unix",
     "Programming Language :: Python",
     "Programming Language :: Python :: 3",
-    "Programming Language :: Python :: 3.9",
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
     "Programming Language :: Python :: 3.12",
@@ -29,12 +28,12 @@ classifiers = [
 ]
 keywords = ["cmasher perceptually uniform sequential colormaps plotting python visualization"]
 
-requires-python = ">=3.9, <4"
+requires-python = ">=3.10, <4"
 
 dependencies = [
     "colorspacious>=1.1.0",
     "matplotlib>=3.5",
-    "numpy>=1.19.5",
+    "numpy>=1.21.2",
 ]
 
 [project.urls]
@@ -116,7 +115,7 @@ convention = "numpy"
 "_backports.py" = ["D"]
 
 [tool.mypy]
-python_version = 3.9
+python_version = '3.10'
 show_error_codes = true
 warn_unused_configs = true
 warn_unused_ignores = true

--- a/requirements/min_constraints.txt
+++ b/requirements/min_constraints.txt
@@ -1,2 +1,0 @@
-scipy<1.11
-contourpy<1.2

--- a/requirements/typecheck.txt
+++ b/requirements/typecheck.txt
@@ -1,3 +1,2 @@
 mypy==1.7.1
-typing-extensions==4.9.0 ; python_version < '3.10'
 pytest==7.4.3

